src/Native/System.Native/pal_networking.cpp

```diff
@@ -379,7 +380,119 @@ static void ConvertHostEntPlatformToPal(HostEntry& hostEntry, hostent& entry)
     }
 }
 
-#if HAVE_GETHOSTBYNAME_R
+#if !HAVE_THREAD_SAFE_GETHOSTBYNAME_AND_GETHOSTBYADDR && (!HAVE_GETHOSTBYNAME_R && !HAVE_GETHOSTBYADDR_R)
```

Comment:
Looked into it more, unfortunately we cannot change this structure as proposed because the second second condition is checking `HAVE_GETNAMEINFO` in the end. This was actually bit tricky to balance in first place because it was failing on FreeBSD, and when fixed against FreeBSD and NetBSD, OSX build failed.
However, we can separate `HAVE_THREAD_SAFE_GET_HOSTBYNAME_AND_GETHOSTBYADDR` for better readability.

Also, these two blocks are consecutive, so we can wrap it inside one !thread-safe condition. But few lines below, we have GetHostByAddrHelper, where we have same three conditions set.